api/agent/drivers/docker/cookie.go

```diff
@@ -306,13 +306,24 @@ func (c *cookie) PullImage(ctx context.Context) error {
 		// TODO need to inspect for hub or network errors and pick; for now, assume
 		// 500 if not a docker error
 		msg := err.Error()
-		code := http.StatusInternalServerError
+		dockerCode := 0
 		if dErr, ok := err.(*docker.Error); ok {
 			msg = dockerMsg(dErr)
-			code = dErr.Status // 401/404
+			dockerCode = dErr.Status // 401/404
 		}
 
-		return models.NewAPIError(code, fmt.Errorf("Failed to pull image '%s': %s", c.task.Image(), msg))
+		finalCode := 0
+
+		switch {
+		case dockerCode == http.StatusGatewayTimeout:
+			finalCode = http.StatusGatewayTimeout
+		case dockerCode >= 400 && dockerCode < 500:
+			finalCode = dockerCode
+		default:
+			finalCode = http.StatusBadGateway
+		}
+
+		return models.NewAPIError(finalCode, fmt.Errorf("Failed to pull image '%s': %s UpstreamStatus=%v", c.task.Image(), msg, dockerCode))
```

Comment:
I don't like this determining of the "final code" this deep in the app. For me, the final code needs to be decided at one point at the front. At this level, we should just be communicating what the problem is, not making decisions about how it should be represented.

Reply:
Updated the code as per the discussion. We only cherry pick 4xx range here.